CoreScriptsRoot/Modules/LoadingScreen3D.lua

```diff
@@ -0,0 +1,355 @@
+					wait(1.5)
+					freeze = false
+				end)
+			else
+				creatorTextPosition = newX
+			end
+		end
+		creatorText.Position = UDim2.new(0, creatorTextPosition, 0, 0)
+	end
+
+	-- if gameNameText.TextFits then
+	-- 	gameNameText.Size = UDim2.new(0.9, 0, 0.1, 0)
+	-- 	gameNameText.Position = UDim2.new(0.05,0,0.65,0)
+	-- 	gameNameText.TextScaled = false
+	-- 	gameNameText.TextWrapped = false
+
+	-- 	spinnerImage.Position = UDim2.new(0.5 - (0.25/2), 0, 0.45 - (0.25/2), 0)
+	-- 	loadingText.Position = UDim2.new(0.5,0,0.2,0)
+	-- else
+	if not gameNameText.TextFits then
+		gameNameText.Size = UDim2.new(0.9, 0, 0.3, 0)
+		gameNameText.Position = UDim2.new(0.05,0,0.5,0)
+		gameNameText.TextScaled = true
+		gameNameText.TextWrapped = true
+
+		spinnerImage.Position = UDim2.new(0.5 - (0.25/2), 0, 0.225, 0)
+		loadingText.Position = UDim2.new(0.5,0,0.15,0)
+	end
+
+end
+
+
+local function CleanUp()
+	if CleanedUp then return end
+	CleanedUp = true
+	surfaceGuiAdorn.Parent = nil
+	RunService:UnbindFromRenderStep("LoadingGui3D")
+end
+
+local function OnGameInfoLoaded()
+	local creatorName = GameInfoProvider:GetCreatorName()
+	if creatorName and creatorName ~= '' then
+		creatorName = string.format("By %s", tostring(creatorName))
+	end
+	gameNameText.Text = GameInfoProvider:GetGameName()
+	creatorText.Text = creatorName
+end
+
+local function OnReplicatingFinished()
+	if game:IsLoaded() or game.Loaded:wait() then
+		if not CleanedUp then
+			FadeElements(loadingSurfaceGui, 1, {}, SECOND_TO_FADE)
+			wait(SECOND_TO_FADE)
+			CleanUp()
+		end
+	end
+end
+
+local function OnDefaultLoadingGuiRemoved()
+	if not CleanedUp then
+		FadeElements(loadingSurfaceGui, 1, {}, SECOND_TO_FADE)
+		wait(SECOND_TO_FADE)
+		CleanUp()
+	end
+end
+
+
+local function UpdateSurfaceGuiPosition()
+	local camera = workspace.CurrentCamera
+	if camera then
+		local cameraCFrame = camera.CFrame
+		local cameraLook = cameraCFrame.lookVector
+		local cameraHorizontalVector = Vector3.new(cameraLook.X, 0, cameraLook.Z).unit
+		local cameraRenderCFrame = camera:GetRenderCFrame()
+
+		surfaceGuiAdorn.CFrame = CFrame.new(cameraRenderCFrame.p + cameraHorizontalVector * GUI_DISTANCE_FROM_CAMERA, cameraRenderCFrame.p)
+	end
+end
+
+do
+	local lastUpdate = tick()
+	RunService:BindToRenderStep("LoadingGui3D", Enum.RenderPriority.Last.Value, function()
+		local now = tick()
+		local delta = now - lastUpdate
+
+		UpdateSurfaceGuiPosition()
+		UpdateLayout(delta)
+
+		local rotation = delta * ROTATIONS_PER_SECOND * 360
+		spinnerRotation = spinnerRotation + rotation
+		spinnerImage.Rotation = spinnerRotation
+
+		lastUpdate = now
+	end)
+	UpdateSurfaceGuiPosition()
+	UpdateLayout()
+end
+
+GameInfoProvider:LoadAssetsAsync()
+if GameInfoProvider:IsReady() then
+	OnGameInfoLoaded()
+end
+-- Disco the following events?
+GameInfoProvider.LoadingFinishedEvent:connect(OnGameInfoLoaded)
+
+
+ReplicatedFirst.FinishedReplicating:connect(OnReplicatingFinished)
```

Comment:
any reason not to check the IsFinishedReplicating first? We can not connect to finishedreplicating in this situation